core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -221,34 +223,52 @@ protected boolean addsDeleteFiles() {
   /** Add a data file to the new snapshot. */
   protected void add(DataFile file) {
     Preconditions.checkNotNull(file, "Invalid data file: null");
-    setDataSpec(file);
-    addedFilesSummary.addedFile(dataSpec(), file);
+    addDataFile(new FileHolder<>(file));
+  }
+
+  /** Add a data file to the new snapshot. */
+  protected void add(DataFile file, long dataSequenceNumber) {
+    Preconditions.checkNotNull(file, "Invalid data file: null");
+    addDataFile(new FileHolder<>(file, dataSequenceNumber));
```

Comment:
The caller can't reliably supply a sequence number. Let's talk about this on Tuesday. I think there's a way that you can update this class to do what you want, but it's going to be a bit more work than what you have here.